File: e2e/cypress-javascript/support/mocks/ApiRouteMock.js
class ApiRouteMock {
  createTask(response) {
    cy.intercept('POST', '**/v1/tasks', { statusCode: 201, body: response });
    return this;
  }

  getTasks(response) {
    cy.intercept('GET', '**/v1/tasks', { statusCode: 200, body: response });
    return this;
  }

  getTask(id, response) {
    cy.intercept('GET', `**/v1/tasks/${id}`, { statusCode: 200, body: response });
    return this;
  }

  deleteTask(id) {
    cy.intercept('DELETE', `**/v1/tasks/${id}`, { statusCode: 204 });
    return this;
  }

  updateTask(id, response) {
    cy.intercept('PUT', `**/v1/tasks/${id}`, { statusCode: 200, body: response });
    return this;
  }

  getIsValid(id, isValid) {
    cy.intercept('GET', `**/v1/tasks/isValid/${id}`, { statusCode: 200, body: isValid });
    return this;
  }
}

module.exports = { ApiRouteMock };

File: e2e/capybara-ruby/support/mocks/api_route_mock.rb
# frozen_string_literal: true

class ApiRouteMock
  JSON_HEADERS = {
    'Content-Type' => 'application/json',
    'Access-Control-Allow-Origin' => '*'
  }.freeze

  CORS_HEADERS = {
    'Access-Control-Allow-Origin' => '*',
    'Access-Control-Allow-Methods' => 'GET, POST, PUT, DELETE, OPTIONS',
    'Access-Control-Allow-Headers' => 'Content-Type'
  }.freeze

  def create_task(response)
    stub_request(:post, %r{/v1/tasks$}, json: response, code: 201)
    self
  end

  def get_tasks(response)
    stub_request(:get, %r{/v1/tasks$}, json: response, code: 200)
    self
  end

  def get_task(id, response)
    stub_request(:get, %r{/v1/tasks/#{Regexp.escape(id)}$}, json: response, code: 200)
    self
  end

  def delete_task(id)
    # Empty 204 body (JSON body can break Chromium fetch). Dup headers — Billy mutates them.
    Billy.proxy.stub(%r{/v1/tasks/#{Regexp.escape(id)}$}, method: 'DELETE').and_return(
      code: 204,
      body: '',
      headers: CORS_HEADERS.dup
    )
    self
  end

  def update_task(id, response)
    stub_request(:put, %r{/v1/tasks/#{Regexp.escape(id)}$}, json: response, code: 200)
    self
  end

  def get_is_valid(id, is_valid)
    stub_request(:get, %r{/v1/tasks/isValid/#{Regexp.escape(id)}$}, json: is_valid, code: 200)
    self
  end

  private

  def stub_request(method, url_pattern, json:, code:)
    Billy.proxy.stub(url_pattern, method: method.to_s.upcase).and_return(
      json: json,
      code: code,
      headers: JSON_HEADERS
    )
  end
end
